bittensor/utils/tokenizer_utils.py

```diff
@@ -0,0 +1,910 @@
+                [[left_0, right_0, substring_0], ...]
+
+        Returns:
+            new_string (:obj:`str`, `required`):
+                New string where replacements were made.
+            new_offsets (:obj:`List[List[Any]]`, `required`):
+                New offsets where replacements are now located
+                [[left_0, right_0], [left_1, right_1], ...]
+    """
+    new_string = ''
+    new_offsets = []
+
+    prev = 0
+    for left, right, substring in offsets:
+        new_string += string[prev:left]  # retain preceding string
+        new_left = len(new_string)  # advance index
+
+        new_string += substring  # add new substring
+        new_right = len(new_string)
+
+        new_offsets += [[new_left, new_right]]  # add offsets
+
+        prev = right  # advance index
+
+    new_string += string[prev:]
+
+    return new_string, new_offsets
+
+
+def get_special_token_pairings(from_tokenizer: PreTrainedTokenizerBase,
+                               to_tokenizer: PreTrainedTokenizerBase) -> Dict[str, str]:
+    r"""
+    Determines a prioritized matching of special token texts between two tokenizers.
+    Purpose is to produce replacement pairs so special token test is correctly represented for target tokenizer.
+        Args:
+            from_tokenizer (:obj:`PreTrainedTokenizerBase`, `required`):
+                From tokenizer.
+            to_tokenizer (:obj:`PreTrainedTokenizerBase`, `required`):
+                To tokenizer.
+
+        Returns:
+            pairings (:obj:`Dict[str, str]`, `required`):
+                Prioritized dictionary of From_special_token_text -> To_special_token_text.
+    """
+    pairings = {}
+
+    # some tokenizers e.g. GPT2 have the same text signifying BOS and EOS, while in other e.g. XGLM they differ
+    # so prioritize EOS token first, since this seems to be the default context separator, e.g. XGLM, GerPT2, GPT2
+    if ('eos_token' in from_tokenizer.special_tokens_map) and ('eos_token' in to_tokenizer.special_tokens_map):
+        pairings[getattr(from_tokenizer, 'eos_token')] = getattr(to_tokenizer, 'eos_token')
+
+    for special_token in from_tokenizer.special_tokens_map:
+        if special_token in to_tokenizer.special_tokens_map:
+            if getattr(from_tokenizer, special_token) not in pairings:  # prevent priority overwrite
+                pairings[getattr(from_tokenizer, special_token)] = getattr(to_tokenizer, special_token)
+
+    return pairings
+
+
+def translate_special_token_text(text_batch: List[str], from_tokenizer: PreTrainedTokenizerBase,
+                                 to_tokenizer: PreTrainedTokenizerBase) -> Tuple[List[str],
+                                                                                 List[List[List[int]]],
+                                                                                 List[List[List[int]]],
+                                                                                 List[List[List[Any]]]]:
```

Comment:
This return type is pretty wild. I think you should create a data structure to encapsulate it in a way that's easier to understand.

Reply:
Noted, may simplify this later on.